### DISCUSSION TITLE
Saving to a temporary file without extension

### DISCUSSION THREAD

**Discussion by jarodium** [libvips/libvips#2641] [Q&A] [answered]
Hello

I would like to ask this discussion group, if any of you have tried to save an image to a php generated temp file?

Here is what I'm trying to do:

1. Create the temporary file with ```php $tmpfname = tempnam(sys_get_temp_dir(), BDPX); ``` with the original file being JPEG and the final destination to be WEBP.
2. Pass that temporary file to another php script.
3. The php script will make a thumbnail 
4. Overwrite the temporary file.

As you all know, the way vips works do not allow overwrite the same path I'm opening, because it is streaming, so I'm bypassing this by using the following code in step 3:

```php
$image = \Jcupitt\Vips\Image::thumbnail($tempFile,$width,['size'=>'VIPS_SIZE_DOWN']);
$buffer = $image->writeToBuffer(".webp");
$image = null;

```
In step 4:

```php
$image =  \Jcupitt\Vips\Image::newFromBuffer($buffer);
$image->writeToFile($tempFile);
```

As expected Vips complains that **Uncaught Jcupitt\Vips\Exception: VipsForeignSave: "/tmp/controloBIi0dC"** is not a known file format. 

Since that there is no options documented in [vips_write_to_file](https://www.libvips.org/API/current/VipsImage.html#vips-image-write-to-file), the only option I am aware is [vips-write-to-target](https://www.libvips.org/API/current/VipsImage.html#vips-image-write-to-target). 

Unfortunately to me, the latest version of [phpvips](https://github.com/libvips/php-vips/blob/master/src/Image.php) does not seem to have a writeToTarget method, so, in order to solve this issue, my only two choices are:

1.To use my own temporary file setup, and this won't do because maintaning a temporary file has it's extra costs and I would very much like to leave it to the operating system.
2. Don't rely on tempnam() function and copy my original file there.

Have you have had an issue where you needed to tackle files from the system's temporary folder?

Best regards
jarodium

**Comment by jarodium** [comment, marked as answer]
Hello all

After some time, I have a solution for my use case.
I was able to sneak a peek at @jcupitt's test scripts and I got some inspiration from them.

I just add ".webp" extension to the temporary file on step 1 ( thus creating a new temporary file ).

Heres some code ( the code sits in a class, so i'll paste the relevant funcionts )

```php

function resize() {
    $image = \Jcupitt\Vips\Image::thumbnail($this->IMAGE,500,["height" => 1920,'size'=>'VIPS_SIZE_DOWN']);
    $this->buffer = $image->writeToBuffer(".webp");
    $image = null;
    return $this;
}

function finalSave() 
    try {
        $image =  \Jcupitt\Vips\Image::newFromBuffer($this->buffer);
        $image->writeToFile($this->IMAGE.".webp");
        return($this->IMAGE.".webp"); 
    }
    catch(Exception $e){}
}

function makeBreakpoints() {
// no implemented yet
}

function crop() {
// no implemented yet
}
```
Imagine that I have a jpeg that has 2081x2081 and I want to save a 1920 width Webp and responsive breakpoints:

1. Initialize this class with the pre-uploaded file ( temp location )
2. Check if user wants to rescale or crop
3. Apply resize or crop methods accordingly
4. Check if user wants to apply breakPoints and make them, if required. This would create a couple more jpeg temp files.
5. Final save image to destination as webp.
6. Final save breakpoints to destination as webp.

In conclusion, I believe for my use case, using buffer should provide effective and since i'm not working with Imagick alonside VIPS i'm not using the writetoMemory way. Maybe in some time in the future I will switch, but at this time, buffer is working for me.

I hope my findings may help someone.

Best regards
jarodium

**Reply by jcupitt** [reply]
This will work, but you are saving as web in stage 1, then in stage 2 loading as webp again and doing a final save. The extra save and load as webp for the intermediate image will be slow. 

If the two stages will be in the same process, you can just pass the `$image` around, there's no need to save and load again.

If the two stages are on different machines and you need to pass the image between them, then I would use something like TIFF or VIPS format (`.tiff` or `.v`). Although they are bigger on disc, they are so much faster I'm sure it'd be worth it. And it sounds like these intermediate images you are using are not so large.

**Reply by jcupitt** [reply]
Tiny benchmark:

```
$ time (vips thumbnail x.jpg x.webp 1920 && vips copy x.webp y.webp)
real	0m0.577s
user	0m0.843s
sys	0m0.152s

$ time (vips thumbnail x.jpg x.v 1920 && vips copy x.v y.webp)
real	0m0.305s
user	0m0.656s
sys	0m0.055s
```

About 2x faster overall with a `.v` intermediate.

It's faster again with no intermediate (this would be the equivalent of just passing the `$image` around):

```
$ time vips thumbnail x.jpg x.webp 1920
real	0m0.274s
user	0m0.538s
sys	0m0.088s
```

**Comment by jcupitt** [comment]
Hi @jarodium,

`writeToFile` uses the output file extension to search for a save operation, then calls that saver with the arguments you supply. 

You can just call the saver directly if you want a specific format, so eg.:

```php
$image->savewebp("filename-with-no-extension");
```

Will write an image to a file without an extension in WEBP format.

**Reply by jarodium** [reply]
I see. Will change my logic to use savewebp.
Once again, thank you for your help :)

**Reply by jarodium** [reply]
Strangely php does not recognize that command. I am using v1.0.9 with php 7.4. 
Maybe it's a newer libvips that is missing from my system...

**Reply by jcupitt** [reply]
Ooops, so sorry, try `webpsave`.

**Reply by jarodium** [reply]
Yes. That dit it.
Thank you
